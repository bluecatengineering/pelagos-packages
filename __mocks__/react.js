module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
