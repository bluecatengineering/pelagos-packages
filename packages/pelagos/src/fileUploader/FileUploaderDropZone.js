import {useCallback, useRef} from 'react';
import PropTypes from 'prop-types';
import {select, t} from '@bluecateng/l10n.macro';

import './FileUploader.less';

const handleDragOver = (event) => {
	event.preventDefault();
	const dt = event.dataTransfer;
	dt.effectAllowed = 'copy';
	dt.dropEffect = 'copy';
};

const handleDragEnter = (event) => {
	const relatedTarget = event.relatedTarget;
	const currentTarget = event.currentTarget;
	if (!relatedTarget || !currentTarget.contains(relatedTarget)) {
		currentTarget.classList.add('FileUploader__dropZone--active');
	}
};

const handleDragLeave = (event) => {
	const relatedTarget = event.relatedTarget;
	const currentTarget = event.currentTarget;
	if (!relatedTarget || !currentTarget.contains(relatedTarget)) {
		currentTarget.classList.remove('FileUploader__dropZone--active');
	}
};

const invokeOnAddFiles = (onAddFiles, liveRef, selectedFiles) => {
	liveRef.current.textContent =
		selectedFiles.length === 1
			? t`File ${selectedFiles[0].name} uploaded`
			: t`Files ${selectedFiles.map((file) => file.name).join(', ')} uploaded`;
	onAddFiles(selectedFiles);
};

const convertToAccept = (types) =>
	types
		? Array.from(new Set(types.flatMap(({accept}) => Object.entries(accept).flatMap(([k, v]) => [k, ...v])))).join(',')
		: null;

/** Drop zone for a file uploader. */
const FileUploaderDropZone = ({id, label, text, types, required, multiple, error, disabled, onAddFiles, ...props}) => {
	const liveRef = useRef();

	const handleClick = useCallback(
		(event) =>
			window.showOpenFilePicker
				? showOpenFilePicker({multiple, types})
						.then((handles) => Promise.all(handles.map((handle) => handle.getFile())))
						.then((selectedFiles) => invokeOnAddFiles(onAddFiles, liveRef, selectedFiles))
						.catch(() => {
							/* ignore all errors */
						})
				: event.currentTarget.nextSibling.click(),
		[multiple, onAddFiles, types]
	);
	const handleDrop = useCallback(
		(event) => {
			event.preventDefault();
			event.currentTarget.classList.remove('FileUploader__dropZone--active');
			const files = event.dataTransfer.files;
			if (files?.length) {
				invokeOnAddFiles(onAddFiles, liveRef, Array.from(files));
			}
		},
		[onAddFiles]
	);
	const handleChange = useCallback(
		(event) => {
			invokeOnAddFiles(onAddFiles, liveRef, Array.from(event.target.files));
		},
		[onAddFiles]
	);
	return (
		<div>
			<div className="sr-only" aria-live="polite" ref={liveRef} />
			<button
				{...props}
				id={id}
				className="FileUploader__dropZone"
				type="button"
				disabled={disabled}
				aria-invalid={error}
				aria-label={select(required, {true: `${label}, required. ${text}`, other: `${label}. ${text}`})}
				onClick={handleClick}
				onDragOver={disabled ? null : handleDragOver}
				onDragEnter={disabled ? null : handleDragEnter}
				onDragLeave={disabled ? null : handleDragLeave}
				onDrop={disabled ? null : handleDrop}>
				{text}
			</button>
			{window.showOpenFilePicker ? null : (
				<input
					type="file"
					className="FileUploader__input"
					multiple={multiple}
					accept={convertToAccept(types)}
					onChange={handleChange}
				/>
			)}
		</div>
	);
};

FileUploaderDropZone.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The label text. */
	label: PropTypes.string,
	/** The text to display. */
	text: PropTypes.string,
	/** The accepted file types. */
	types: PropTypes.array,
	/** Whether the field is required. */
	required: PropTypes.bool,
	/** Whether multiple files are allowed. */
	multiple: PropTypes.bool,
	/** Whether the component is in error. */
	error: PropTypes.bool,
	/** Whether the field is disabled. */
	disabled: PropTypes.bool,
	/** Function invoked when files are added. */
	onAddFiles: PropTypes.func,
};

export default FileUploaderDropZone;
