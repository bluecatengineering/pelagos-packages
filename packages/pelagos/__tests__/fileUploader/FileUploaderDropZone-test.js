import {useRef} from 'react';
import {shallow} from 'enzyme';

import FileUploaderDropZone from '../../src/fileUploader/FileUploaderDropZone';

jest.unmock('../../src/fileUploader/FileUploaderDropZone');

const types = [
	{description: 'Foo files', accept: {'text/x-foo': ['.foo'], 'application/x-foo': ['.foo']}},
	{description: 'Bar files', accept: {'text/x-bar': ['.bar1', '.bar2'], 'text/x-baz': ['.baz']}},
];

describe('FileUploaderDropZone', () => {
	afterEach(() => delete global.showOpenFilePicker);

	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(<FileUploaderDropZone label="Test" text="Test drop zone" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when optional properties are set', () => {
			const wrapper = shallow(
				<FileUploaderDropZone
					id="test"
					label="Test"
					text="Test drop zone"
					types={types}
					required
					multiple
					error
					disabled
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when showOpenFilePicker is present', () => {
			global.showOpenFilePicker = true;
			const wrapper = shallow(<FileUploaderDropZone label="Test" text="Test drop zone" types={types} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onAddFiles when a file is dropped', () => {
			const onAddFiles = jest.fn();
			const preventDefault = jest.fn();
			const remove = jest.fn();
			const name = 'file.txt';
			const files = [{name}];
			const live = {};
			useRef.mockReturnValueOnce({current: live});
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} onAddFiles={onAddFiles} />);

			wrapper.find('#test').prop('onDrop')({
				preventDefault,
				currentTarget: {classList: {remove}},
				dataTransfer: {files},
			});
			expect(preventDefault.mock.calls).toEqual([[]]);
			expect(remove.mock.calls).toEqual([['FileUploader__dropZone--active']]);
			expect(onAddFiles.mock.calls).toEqual([[files]]);
			expect(live.textContent).toBe('File file.txt uploaded');
		});

		it('clicks the input element when the drop zone is clicked', () => {
			const click = jest.fn();
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} />);
			wrapper.find('#test').simulate('click', {currentTarget: {nextSibling: {click}}});
			expect(click.mock.calls).toEqual([[]]);
		});

		it('calls onAddFiles when the drop zone is clicked and showOpenFilePicker is available', () => {
			const onAddFiles = jest.fn();
			const live = {};
			useRef.mockReturnValueOnce({current: live});
			global.showOpenFilePicker = jest
				.fn()
				.mockResolvedValue([
					{getFile: jest.fn().mockResolvedValue({name: 'bar.txt'})},
					{getFile: jest.fn().mockResolvedValue({name: 'baz.txt'})},
				]);
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} multiple onAddFiles={onAddFiles} />);
			return wrapper
				.find('#test')
				.prop('onClick')()
				.then(() => {
					expect(showOpenFilePicker.mock.calls).toEqual([[{multiple: true, types: []}]]);
					expect(onAddFiles.mock.calls).toEqual([[[{name: 'bar.txt'}, {name: 'baz.txt'}]]]);
					expect(live.textContent).toBe('Files bar.txt, baz.txt uploaded');
				});
		});

		it('ignores errors when showOpenFilePicker fails', () => {
			global.showOpenFilePicker = jest.fn().mockRejectedValue(new Error());
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} />);
			return wrapper
				.find('#test')
				.prop('onClick')()
				.then(() => {
					expect(showOpenFilePicker.mock.calls).toEqual([[{types: []}]]);
				});
		});

		it('calls onAddFiles when a file is selected', () => {
			const onAddFiles = jest.fn();
			const name = 'file.txt';
			const files = [{name}];
			const live = {};
			useRef.mockReturnValueOnce({current: live});
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} multiple onAddFiles={onAddFiles} />);

			wrapper.find('input').simulate('change', {target: {files}});
			expect(onAddFiles.mock.calls).toEqual([[files]]);
			expect(live.textContent).toBe('File file.txt uploaded');
		});

		it('calls preventDefault on dragover', () => {
			const preventDefault = jest.fn();
			const dataTransfer = {};
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} />);
			wrapper.find('#test').prop('onDragOver')({preventDefault, dataTransfer});
			expect(preventDefault.mock.calls).toEqual([[]]);
			expect(dataTransfer).toEqual({effectAllowed: 'copy', dropEffect: 'copy'});
		});

		it('adds a class on dragenter', () => {
			const relatedTarget = {};
			const contains = jest.fn();
			const add = jest.fn();
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} />);
			wrapper.find('#test').prop('onDragEnter')({relatedTarget, currentTarget: {contains, classList: {add}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(add.mock.calls).toEqual([['FileUploader__dropZone--active']]);
		});

		it('does not add a class on dragenter when contains returns true', () => {
			const relatedTarget = {};
			const contains = jest.fn().mockReturnValue(true);
			const add = jest.fn();
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} />);
			wrapper.find('#test').prop('onDragEnter')({relatedTarget, currentTarget: {contains, classList: {add}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(add.mock.calls).toEqual([]);
		});

		it('removes a class on dragleave', () => {
			const relatedTarget = {};
			const contains = jest.fn();
			const remove = jest.fn();
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} />);
			wrapper.find('#test').prop('onDragLeave')({relatedTarget, currentTarget: {contains, classList: {remove}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(remove.mock.calls).toEqual([['FileUploader__dropZone--active']]);
		});

		it('does not remove a class on dragleave when contains returns true', () => {
			const relatedTarget = {};
			const contains = jest.fn().mockReturnValue(true);
			const remove = jest.fn();
			const wrapper = shallow(<FileUploaderDropZone id="test" types={[]} />);
			wrapper.find('#test').prop('onDragLeave')({relatedTarget, currentTarget: {contains, classList: {remove}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(remove.mock.calls).toEqual([]);
		});
	});
});
