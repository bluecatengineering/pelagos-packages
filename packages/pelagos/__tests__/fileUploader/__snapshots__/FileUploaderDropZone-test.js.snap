// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploaderDropZone rendering renders expected elements 1`] = `
<div>
  <div
    aria-live="polite"
    className="sr-only"
  />
  <button
    aria-label="Test. Test drop zone"
    className="FileUploader__dropZone"
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    type="button"
  >
    Test drop zone
  </button>
  <input
    accept={null}
    className="FileUploader__input"
    onChange={[Function]}
    type="file"
  />
</div>
`;

exports[`FileUploaderDropZone rendering renders expected elements when optional properties are set 1`] = `
<div>
  <div
    aria-live="polite"
    className="sr-only"
  />
  <button
    aria-invalid={true}
    aria-label="Test, required. Test drop zone"
    className="FileUploader__dropZone"
    disabled={true}
    id="test"
    onClick={[Function]}
    onDragEnter={null}
    onDragLeave={null}
    onDragOver={null}
    onDrop={null}
    type="button"
  >
    Test drop zone
  </button>
  <input
    accept="text/x-foo,.foo,application/x-foo,text/x-bar,.bar1,.bar2,text/x-baz,.baz"
    className="FileUploader__input"
    multiple={true}
    onChange={[Function]}
    type="file"
  />
</div>
`;

exports[`FileUploaderDropZone rendering renders expected elements when showOpenFilePicker is present 1`] = `
<div>
  <div
    aria-live="polite"
    className="sr-only"
  />
  <button
    aria-label="Test. Test drop zone"
    className="FileUploader__dropZone"
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    type="button"
  >
    Test drop zone
  </button>
</div>
`;
